blog/content/post/2019-05-15-building-bareflank.md

```diff
@@ -0,0 +1,345 @@
+file is searched for in standard UEFI search for bootloaders. Because of that,
+it is difficult to start Bareflank and then the real bootloader automatically.
+There are basically 3 options for starting Bareflank. All files mentioned
+below (except `bareflank.efi`) are located in `\EFI\BOOT\` directory on the USB
+drive.
+
+###### 1st option
+
+Unless UEFI Shell can be forced to start despite a valid `bootx64.efi` file,
+the most automatic option is to modify the Bareflank source to start a different
+file. The line responsible for that is located in `load_start_vm()`, in
+`hypervisor/bfdriver/src/platform/efi/entry.c`. Change it to point to
+`grubx64.efi` to start GRUB directly and copy `bareflank.efi` to `bootx64.efi`.
+
+###### 2nd option
+
+Another option is to just remove the original `bootx64.efi`, put the
+`bareflank.efi` in the root directory of the USB drive and start the platform.
+It should fall back to the shell, then `bareflank.efi` can be started manually.
+It will fail to load `bootx64.efi` and return to the shell, where the
+`grubx64.efi` bootloader file can be started.
+
+###### 3rd option
+
+The third option is to replace the original `bootx64.efi` with [UEFI Shell](https://github.com/tianocore/edk2/tree/master/ShellBinPkg/UefiShell/X64).
+This option is useful when there is no built-in shell available, as is the case
+with most PC firmwares. In this case Bareflank will start another instance of
+UEFI Shell, but it's OK, the shell memory is not reserved and can be reused
+later when the OS starts its own memory manager.
+
+The first option is fully automatic, while the next two give more room for e.g.
+comparing memory maps before and after starting Bareflank (try it). They can
+also print error messages when something breaks along the way. In the rest of
+this post I'll be showing output as it is visible with option 2.
+
+![Bareflank startup](/img/bareflank_start.png)
+
+Bareflank even can print strings in colours (like, all 8 of them!), as long as
+the client handles [ANSI escape codes](https://en.wikipedia.org/wiki/ANSI_escape_code).
+The `DEBUG: host os is now in a vm` lines are visible only on the serial output.
+UEFI overwrites Bareflank output on the attached image because it is not aware
+that something else is printing through UART.
+
+As expected, the Bareflank entry point failed to start a bootloader. It didn't
+stop the VMX operation, so the UEFI is now running in VMX non-root mode, as
+will be the OS that we are going to start now:
+
+```
+FS0:\> EFI\BOOT\grubx64.efi
+```
+
+Remember to use backslashes, tab-completion is somewhat buggy otherwise. GRUB
+may not be printed on the serial output, but the OS should start automatically
+after 5 seconds, plus some more before Ubuntu starts printing anything. After
+it loads you can log in using default user `ubuntu`, or you can just use this
+system with GUI. It should behave more or less just like it does without any
+hypervisor.
```

Comment:
@krystian-hebel you didn't mentioned on what platform you running whole experiment.

Reply:
https://github.com/3mdeb/news-and-ideas/blame/5ceb7eb8046cb3a06f9a2cde7cfd5a24b8b125d0/blog/content/post/2019-05-15-building-bareflank.md#L38

I guess I could move it to separate paragraph.

Reply:
Damn, how I missed that.